### ISSUE TITLE
weird af memory corruption bug

### ISSUE THREAD

**Issue by dmcdougall** [libqueso/queso#582]
@pbauman reported this to me privately.  I'm dumping a recreate here.

Application:
```c
#include <queso/Environment.h>
#include <queso/VectorSpace.h>
#include <queso/GslVector.h>
#include <queso/GslMatrix.h>
#include <queso/BoxSubset.h>
#include <queso/GaussianLikelihoodScalarCovariance.h>

template<typename VectorType = QUESO::GslVector, typename MatrixType = QUESO::GslMatrix>
class Likelihood : public QUESO::GaussianLikelihoodScalarCovariance<VectorType,MatrixType>
{
public:
  Likelihood( const QUESO::VectorSet<VectorType,MatrixType> & domain,
              const VectorType & observations,
              double variance )
    : QUESO::GaussianLikelihoodScalarCovariance<VectorType,MatrixType>("", domain, observations, variance)
  {
    queso_assert_equal_to(this->m_observations.sizeGlobal(), 1);
  }

  virtual void 	evaluateModel (const VectorType & domainVector,
                               const VectorType * /*domainDirection*/,
                               VectorType & modelOutput,
                               VectorType * /*gradVector*/,
                               MatrixType * /*hessianMatrix*/,
                               VectorType * /*hessianEffect*/) const
  {
    queso_assert_equal_to( modelOutput.sizeGlobal(), this->m_observations.sizeGlobal() );
  }
};

int main(int argc, char ** argv)
{
  MPI_Init(&argc, &argv);

  QUESO::FullEnvironment env(argv[1], "", NULL);
  QUESO::VectorSpace<> param_space( env, "", 2, NULL );
  QUESO::VectorSpace<> data_space( env, "", 1, NULL );

  QUESO::GslVector obs( data_space.zeroVector() );
  obs[0] = 4603.50;

  QUESO::GslVector min_vals( param_space.zeroVector() );
  QUESO::GslVector max_vals( param_space.zeroVector() );

  QUESO::BoxSubset<> param_domain( "", param_space, min_vals, max_vals );

  double variance = 10.0*10.0;

  Likelihood<> likelihood( param_domain, obs, variance );

  MPI_Finalize();

  return 0;
}
```

Input file:
```
###############################################
# UQ Environment
###############################################
#env_help                = anything
env_numSubEnvironments   = 1
env_subDisplayFileName   = outputData/display
env_subDisplayAllowAll   = 0
env_subDisplayAllowedSet = 0
env_displayVerbosity     = 2
env_syncVerbosity        = 0
env_seed                 = 0

###############################################
# Statistical inverse problem (ip)
###############################################
#ip_help                 = anything
ip_computeSolution      = 1
ip_dataOutputFileName   = outputData/sipOutput
ip_dataOutputAllowedSet = 0

###############################################
# 'ip_': information for Metropolis-Hastings algorithm
###############################################
#ip_mh_help                 = anything
ip_mh_dataOutputFileName   = outputData/sipOutput
ip_mh_dataOutputAllowedSet = 0 1
ip_mh_dataOutputFileType = h5 #txt

ip_mh_rawChain_dataInputFileName    = .
ip_mh_rawChain_size                 = 10000
ip_mh_rawChain_generateExtra        = 0
ip_mh_rawChain_displayPeriod        = 50000
ip_mh_rawChain_measureRunTimes      = 1
ip_mh_rawChain_dataOutputFileName   = outputData/ip_raw_chain
ip_mh_rawChain_dataOutputAllowedSet = 0 1
ip_mh_rawChain_computeStats         = 1

ip_mh_displayCandidates             = 0
ip_mh_putOutOfBoundsInChain         = 0
ip_mh_tk_useLocalHessian            = 0
ip_mh_tk_useNewtonComponent         = 1
ip_mh_dr_maxNumExtraStages          = 1
ip_mh_dr_listOfScalesForExtraStages = 5.
ip_mh_am_initialNonAdaptInterval    = 0
ip_mh_am_adaptInterval              = 100
ip_mh_am_eta                        = 1.92
ip_mh_am_epsilon                    = 1.e-5

ip_mh_filteredChain_generate             = 0
```

**Comment by dmcdougall** [comment]
This application segfaults.  Apparently `m_observations` is nonsense in the derived class after the base class `LikelihoodBase` constructor has finished executing.

**Comment by pbauman** [comment]
Two notes: First, in your pasted program, change the assert in `Likelihood` constructor to check for size 1 instead of 7. Second, this only fails for me with current dev, 0.57.0 release is fine. I will try and find time to `git bisect` this sometime this week.

**Comment by dmcdougall** [comment]
Oh yeah, sorry.  I'll edit the assert.

**Comment by pbauman** [comment]
It gets better. The reduced test program fails for me on commits before the 0.57.0 release, so I have yet to find a good commit to run the test case on. My intent was to `git bisect run` instead of doing dishes. Now, I still have dirty dishes and `git bisect run` isn't running.

**Comment by pbauman** [comment]
I just tried this with diagonal covariance instead of scalar and it's fine...

**Comment by pbauman** [comment]
And have tried both scalar and diagonal with the libMesh WTF-dbg flags, including GLIBCXX_DEBUG (includes building QUESO that way) and still the same result: scalar case dies, diagonal case is fine.

Also, there's a shit load of warnings with all those dbg flags. We need to clean all that shit up.

**Comment by pbauman** [comment]
Hmm, OK, I can get a good result with the hash for the 0.57.0 this morning. I might've been using the wrong path last night. I'll kick off a git bisect run and see what happens.

**Comment by pbauman** [comment]
And it gets weirder.

`git bisect run` fingered 00477bd as the offending commit. I manually checked the commit before and after and this seems to be correct. Looks like this is where program options became opt-in. Let me try opt-ing in and see what happens.

**Comment by pbauman** [comment]
Opting into program options gets test code working in both offending commit and dev HEAD. Will try and also confirm on my laptop. I'm guessing we have NULL dereferencing or some other undefined behavior happening in the parsing of the Environment...

**Comment by pbauman** [comment]
Also, we should capture this test in the CppUnit test suite, I think, since this is tripping whatever is happening for us.

**Comment by dmcdougall** [comment]
I think the fact it doesn't segfault with the diagonal case is a red herring; I have an even simpler test case that directly subclasses `LikelihoodBase` and can recreate a segfault.

Thanks for the hard work finding the offending commit.  I wonder if we're still trying to use boost somewhere we shouldn't be (or not handling the GetPot object appropriately).

**Comment by pbauman** [comment]
These were the CXXFLAGS I was using on my workstation, but the result seems to be independent of these. But these will expose the aforementioned warnings. Note the last two flags will require you to have built CppUnit and Boost with those flags. Otherwise, just remove them.

<pre>
CXXFLAGS="-O0 -felide-constructors -g -pedantic -W -Wall -Wextra -Wno-long-long -Wunused -Wpointer-arith -Wformat -Wparentheses -Woverloaded-virtual -DDEBUG -D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC"
</pre>

**Comment by pbauman** [comment]
Confirmed, same behavior on my laptop. So this seems to definitely be something with parsing not using Boost. :(

**Comment by dmcdougall** [comment]
I can confirm with I configure with boost everything works.

**Comment by pbauman** [comment]
Just to be clear, with our without `--enable-boost-program-options`?

**Comment by dmcdougall** [comment]
`./configure --without-boost` produces a segfaulting binary.  `./configure --with-boost=/path/to/boost` produces a non-segfaulting binary.

**Comment by pbauman** [comment]
So you're not using program options then? Because program options was opt-in regardless of `--with-boost`, I thought? For me, I have to use boost (separate issue), but it was the enabling vs. not enabling of program options that triggered the failure. *Not* using program options triggers the failure for me.

**Comment by dmcdougall** [comment]
`--without-boost` implies without boost program options.  I can try with the `--without-boost-program-options`.

**Comment by pbauman** [comment]
> --without-boost implies without boost program options.

Right.

> I can try with the --without-boost-program-options flag as well.

The interesting case will be `--with-boost=<> --enable-boost-program-options` vs. just `--with-boost=<>` since the latter, at the dev HEAD should not use program options, AFAICT. The former case has the test passing for me while the latter does not.

**Comment by dmcdougall** [comment]
Oh, right.  I'll double-check.

**Comment by dmcdougall** [comment]
Alright, something weird is up.  I can't get queso to link to boost with `--with-boost=<> --enable-boost-program-options`.  The compile succeeds, but the link stage fails and complains about missing boost symbols.  I'm too tired to dig more, so I'll look at it tomorrow.

**Comment by pbauman** [comment]
This happened on my laptop too, but not on my workstation. I'm tied up all day today and most of tomorrow, but after that I'll work on the build system and fix those inconsistencies.

**Comment by dmcdougall** [comment]
This is happening on my workstation.  Ok I'll let you deal with that.  At least I have a segfaulting binary I can recreate.

**Comment by dmcdougall** [comment]
@pbauman To summarise our phone conversation:

[These lines contain a macro](https://github.com/libqueso/queso/blob/e77509d12686d4c6b7f285a1953beaec2f901973/src/core/inc/Environment.h#L419-L422) not processed through queso_config.h and exists in the installed header on the user's system.  So if the user is not using boost program options, then the user also needs to pass `-DDISABLE_BOOST_PROGRAM_OPTIONS` when they compile their statistical application.  If the user doesn't do this, the environment object is a different size when compiling the user's statistical application.

I haven't confirmed this is the problem, but it's a good bet.  I'll report back shortly.

**Comment by dmcdougall** [comment]
Confirmed.  Passing `-DDISABLE_BOOST_PROGRAM_OPTIONS` for my statistical application produces a non-segfaulting binary.

**Comment by dmcdougall** [comment]
Should be fixed in #586.

**Comment by dmcdougall** [comment]
Fixed in #586.  It was also merged to dev in commit 12cdc9f.
